client/main/infobox.js

```diff
@@ -28,39 +29,151 @@ class InfoBox extends HTMLElement {
                     </slot>
                 </div>
                 <div class="remote" style="display: none">
                 </div>
+                <div class="close">
+                    <svg xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns="http://www.w3.org/2000/svg" height="10" width="10" version="1.1" xmlns:cc="http://creativecommons.org/ns#" xmlns:dc="http://purl.org/dc/elements/1.1/" viewBox="0 0 10 10">
+                        <g transform="translate(0 -1042.4)" stroke="#777777" stroke-linecap="round" stroke-width="1.2" fill="none">
+                            <path d="m0.48949 1051.9c8.9785-8.9782 9.021-9.0209 9.021-9.0209"/>
+                            <path d="m9.5105 1051.9c-8.9785-8.9782-9.021-9.0209-9.021-9.0209"/>
+                        </g>
+                    </svg>
+                <div>
             </div>`;
 
         this._body = this._root.querySelector('.body');
         this._local = this._body.querySelector('.local');
         this._remote = this._body.querySelector('.remote');
-        this._heading = this._body.querySelector('h1');
+        this._heading = this._body.querySelector('.heading .title');
         this._content = this._body.querySelector('.content');
-        this._button = this._body.querySelector('button');
+        this._button = this._body.querySelector('.button-box button');
+        this._close = this._body.querySelector('.close');
+
+        this._close.addEventListener('click', () => this.hide());
+        this._host.addEventListener('click', (event) => this._backgroundClicked(event));
+        this._host.addEventListener('keydown', (event) => this._onKeyDown(event));
 
         this._button.addEventListener('click', () => this.clicked());
         window.addEventListener('message', (event) => this.onMessage(event));
     }
 
-    setup(info) {
-        this._host.setAttribute('title', info.title || '');
-        this._host.setAttribute('message', info.message || '');
-        this._host.setAttribute('message-src', info['message-src'] || '');
-        this._host.setAttribute('status', info.status || '');
+    setup(info, params) {
+        if (params === undefined)
+            params = { };
+
+        if (this._displayInfo)
+            this.hide();
+
+        this._visible = true;
+        this._processEnterKey = false;
+
+        this._displayInfo = info;
+        if (info['message-src']) {
+            if (params.cancelable === undefined)
+                params.cancelable = false;
+
+            this._local.classList.remove('external');
+            this._body.style.height = `300px`;
+            this._body.style.width = `300px`;
+            this._host.setAttribute('message-src', info['message-src'] || '');
+            this._host.style.display = null;
+            keyboardJS.pause();
+        }
+        else {
+            if (params.cancelable === undefined)
+                params.cancelable = true;
```

Comment:
in some places you've called this 'cancel', in other places 'close', can we make these consistent?